### ISSUE TITLE
Table of coordinate changes for a given gene in the gene page

### ISSUE THREAD

**Issue by manulera** [pombase/website#1993]
The idea would be to have a table where the changes in gene structure for a given gene are listed. This can be obtained from the tsv files:

* `results/all_coordinate_changes_file_comments.tsv`: changes from svn
* `results/pre_svn_coordinate_changes_file_comments.tsv`: changes from pre-svn ftp folders

They are in the genome_changelog repo in this folder:

https://github.com/pombase/genome_changelog/tree/master/results

This is an example of grepping `SPAC1565.08` (cdc48) on both files.

```tsv
revision	user	date	chromosome	systematic_id	primary_name	feature_type	added_or_removed	value	db_xref	pombase_reference	pombase_comments
8650	kmr44	2023-01-12	I	SPAC1486.05		5'UTR	added	3196965..3197027	PMID:30566651		
8650	kmr44	2023-01-12	I	SPAC1486.05		5'UTR	removed	3196998..3197027	PMID:18641648,PMID:20118936		
3985	vw253	2017-04-12	I	SPAC1486.05	nup189	CDS	added	join(3197028..3197529,3197529..3202450)		PMID:26615217; pers. comm. Li-Lin Du	Frameshifted by Chr_I:3197528!A→AG
3985	vw253	2017-04-12	I	SPAC1486.05	nup189	CDS	removed	join(3197028..3197525,3197528..3202450)			
2917	vw253	2015-09-01	I	SPAC1486.05	nup189	CDS	added	join(3197028..3197525,3197528..3202450)			
2917	vw253	2015-09-01	I	SPAC1486.05	nup189	CDS	removed	join(3197028..3197528,3197528..3202450)			
2716	vw253	2015-04-14	I	SPAC1486.05	nup189	CDS	added	join(3197028..3197528,3197528..3202450)			
2716	vw253	2015-04-14	I	SPAC1486.05	nup189	CDS	removed	join(3197028..3197443,3197530..3202450)		pers. comm. Y. Hiraoka	
20110324	-	2011-03-24	I	SPAC1486.05	nup189	3'UTR	removed	3202451..3202485			
20110204	-	2011-02-04	I	SPAC1486.05		5'UTR	added	3196998..3197027			
20050705	-	2005-07-05	I	SPAC1486.05		3'UTR	added	3203351..3203385			
```

* `db_xref` lists additions / removals of db_xref qualifiers associated with a feature that occurred in the same revision.
* `pombase_reference` and `pombase_comments` come from the curation repo `data_files/gene-coordinate-change-data.tsv` matched to the closest date.

> **NOTE**: This table contains changes in gene structure, not coordinates. As in, if the genome sequence changed, and the gene coordinates were shifted so that it sequence remained the same, the change is not listed.

As I said on the call the other day, the "unique identifier" of a feature is the combination of `systematic_id` + `feature_type`, so if the feature type is changed, the change is listed as below:

```tsv
revision	user	date	chromosome	systematic_id	primary_name	feature_type	added_or_removed	value
8658	kmr44	2023-01-16	I	SPNCRNA.2013		lncRNA	added	complement(18768..19648)
8658	kmr44	2023-01-16	I	SPNCRNA.2013		ncRNA	removed	complement(18768..19648)
7957	kmr44	2022-01-12	I	SPNCRNA.2013		ncRNA	added	complement(18768..19648)
```

Maybe then it makes no sense to show those changes, since they are not really gene structure changes. Equivalent files omitting those kinds of change can be found in:

* `results/all_coordinate_changes_file_comments_no_type_change.tsv`: changes from svn
* `results/pre_svn_coordinate_changes_file_comments_no_type_change.tsv`: changes from pre-svn ftp folders

For the same id, you get only this single line from those files:

```tsv
7957    kmr44   2022-01-12      I       SPNCRNA.2013            ncRNA   added   complement(18768..19648)
```

So my suggestion is to use the files that end with `_no_type_change.tsv`.

The relevant columns to be included (I think) are:

* date
* chromosome (sometimes there has been a feature with the same id in another chromosome, often by mistake)
* systematic_id (can be omitted in gene page)
* primary_name (can be omitted in gene page)
* feature_type
* added_or_removed
* value
* db_xref
* pombase_reference (could be merged with db_xref)
* pombase_comments

**Comment by manulera** [comment]
One more thing:

The changelog of coordinates registers when features are added or removed, as well as when they change.

However, the changelog of qualifiers (where I list for example the changes in db_xref) only registers changes (otherwise it would be a massive file), so no db_xref comment is linked to the creation of a feature or the removal. In principle those qualifiers are still there, so accessible from the gene feature in Jbrowse, or if they were changed at the same time as the gene structure (pressumably because the removed PMID descibed the old structure), they would appear in the rows of the table.

* [ ] To capture this, I could at least link the comments in this file in a future version:

https://github.com/pombase/genome_changelog/blob/master/gene_changes_comments_and_pmids/new-gene-data.tsv

https://github.com/pombase/genome_changelog/blob/master/gene_changes_comments_and_pmids/removed-gene-data.tsv

**Comment by kimrutherford** [comment]
On the genes pages what columns should the table have?  Is this enough?:

  - date
  - description of change
    - do we want to display when the gene was added to PomBase when with "genome project" as the default?
  - db_xref/pombase_reference
  - pombase_comments

**Comment by ValWood** [comment]
Sounds good. @manulera  any other suggestions?

**Comment by manulera** [comment]
Hi @kimrutherford, thinking about what we discussed yesterday I have made another file that maybe makes things a bit easier.

What I did was to take the old `*_coordinate_changes_file_comments_no_type_change` and create a new file where only unique removed coordinate values are shown (should contain all non-current values only once), and where comments and db_xref are combined comma separated. For the db_xref column, the added and removed values are in separated columns. I think this might be more convenient. Sorry for not having thought about it before, and maybe you have already done something on your side already, but just in case it helps.

* script: https://github.com/pombase/genome_changelog/blob/master/make_all_previous_coords_file.py
* result: https://github.com/pombase/genome_changelog/blob/master/results/all_previous_coords.tsv

It contains both svn and pre-svn coordinates.

**Comment by kimrutherford** [comment]
Thanks Manu.

Is that new file missing a row for SPBC21C3.07c?

In the older file we have:

```
20071219        -       2007-12-19      II      SPBC21C3.07c            CDS     added   complement(join(3807920..3808607,3808665..3808795))                     Solexa intron data; removed final exon, not supported by data
20071219        -       2007-12-19      II      SPBC21C3.07c            CDS     removed complement(join(3807714..3807783,3807963..3808607,3808665..3808795))
```

but there is no row with the date 2007-12-19 in the new file.

**Comment by manulera** [comment]
So the file keeps only unique values of removed rows, and the latest date where that coordinate existed

This:
```
8650	kmr44	2023-01-12	II	SPBC21C3.07c		5'UTR	added	complement(3808796..3808853)	PMID:30566651		
8650	kmr44	2023-01-12	II	SPBC21C3.07c		5'UTR	removed	complement(3808796..3808991)	PMID:21511999		
20110324	-	2011-03-24	II	SPBC21C3.07c		3'UTR	added	complement(3807091..3807713)			
20110324	-	2011-03-24	II	SPBC21C3.07c		5'UTR	added	complement(3808796..3808991)			
20080222	-	2008-02-22	II	SPBC21C3.07c		CDS	removed	complement(join(3807827..3807891,3807946..3808607,3808665..3808795))			
20080222	-	2008-02-22	II	SPBC21C3.07c		CDS	removed	complement(join(3807920..3808607,3808665..3808795))			Solexa transread data; added two 3’ exons
20080123	-	2008-01-23	II	SPBC21C3.07c		CDS	added	complement(join(3807714..3807783,3807831..3807891,3807946..3808607,3808665..3808795))			
20080123	-	2008-01-23	II	SPBC21C3.07c		CDS	added	complement(join(3807827..3807891,3807946..3808607,3808665..3808795))			
20071219	-	2007-12-19	II	SPBC21C3.07c		CDS	added	complement(join(3807920..3808607,3808665..3808795))			Solexa intron data; removed final exon, not supported by data
20071219	-	2007-12-19	II	SPBC21C3.07c		CDS	removed	complement(join(3807714..3807783,3807963..3808607,3808665..3808795))			
```

Becomes this:

```
II      SPBC21C3.07c            5'UTR   complement(3808796..3808991)    2023-01-12      PMID:21511999
II      SPBC21C3.07c            CDS     complement(join(3807827..3807891,3807946..3808607,3808665..3808795))    2008-02-22
II      SPBC21C3.07c            CDS     complement(join(3807920..3808607,3808665..3808795))     2008-02-22                              Solexa transread data; added two 3’ exons
```

I thought comments from pombase were always matched to the "removed" comment, but I can see it's not the case here. Not sure why that is, I have to check. Perhaps it would be best to link the comment to both removed and added?

**Comment by manulera** [comment]
Sorry, there was indeed a mistake, and `complement(join(3807714..3807783,3807963..3808607,3808665..3808795))` was missing (coordinates are so similar that I could not tell). I was doing an inner join instead of a left join, so that the changes that did not appear as added were lost. Now it should be fixed. Sorry about that. I will push a fix shortly. I will also include the pombase comments.

**Comment by manulera** [comment]
Hi Kim, it's fixed now, I think. I have kept the comments associated with the addition and removal of a given gene, so it looks like this:

| chromosome | systematic_id | primary_name | feature_type | value                                                                | date       | db_xref_removed | db_xref_added | pombase_reference_removed | pombase_reference_added | pombase_comments_removed                                      | pombase_comments_added                                        |
|------------|---------------|--------------|--------------|----------------------------------------------------------------------|------------|-----------------|---------------|---------------------------|-------------------------|---------------------------------------------------------------|---------------------------------------------------------------|
| II         | SPBC21C3.07c  |              | 5'UTR        | complement(3808796..3808991)                                         | 2023-01-12 | PMID:21511999   |               |                           |                         |                                                               |                                                               |
| II         | SPBC21C3.07c  |              | CDS          | complement(join(3807714..3807783,3807963..3808607,3808665..3808795)) | 2007-12-19 |                 |               |                           |                         | Solexa intron data; removed final exon, not supported by data |                                                               |
| II         | SPBC21C3.07c  |              | CDS          | complement(join(3807827..3807891,3807946..3808607,3808665..3808795)) | 2008-02-22 |                 |               |                           |                         |                                                               | Solexa transread data; added two 3’ exons                     |
| II         | SPBC21C3.07c  |              | CDS          | complement(join(3807920..3808607,3808665..3808795))                  | 2008-02-22 |                 |               |                           |                         | Solexa transread data; added two 3’ exons                     | Solexa intron data; removed final exon, not supported by data |

**Comment by kimrutherford** [comment]
Thanks Manu.  That file is much easier to parse and the gene history tables are looking better now.

Here's an example: https://desktop.kmr.nz/gene/SPBC21C3.07c

**Comment by kimrutherford** [comment]
The PMID IDs in the references are links now.  The other IDs will take longer.  Here's another example:

https://desktop.kmr.nz/gene/SPSNRNA.06

![image](https://user-images.githubusercontent.com/90474/216249505-437b798c-ecbb-472c-96da-8215319584a7.png)

**Comment by ValWood** [comment]
If the comments ever need editing to make them more readable , where would we do that?

**Comment by manulera** [comment]
@ValWood the files that end with data.tsv in this folder:

https://github.com/pombase/genome_changelog/tree/master/gene_changes_comments_and_pmids

**Comment by kimrutherford** [comment]
I forgot about this issue.  The table is available on the dev site:
http://dev.pombase.kmr.nz/gene/SPCC622.17

There are few things to chat about next time we have a call.

**Comment by manulera** [comment]
Looks good to me!

**Comment by kimrutherford** [comment]
Change header:
`Date -> Used until`

Add a new row with "now" for the date and the current structure.

**Comment by manulera** [comment]
I think if we use "used until" we can remove the "previous" of the second header

**Comment by kimrutherford** [comment]
I've made the changes that we talked about including adding the gene history to the left hand menu.  Is there anything else to do?  Do we need a information/help icon for the gene history sections?

**Comment by manulera** [comment]
Maybe I would rename the block to "Gene Structure History", to be a bit more specific.

We can discuss this when we speak, and maybe we don't want to go into this, but there is an extra thing that we could do. The displayed coordinates refer to the genome sequence that existed at that moment in the contig file. This means that if the genome sequence since then, especially with insertions / deletions of nucleotides, using that feature location on the current genome sequence might not return the right sequence. The log of where the genome sequence changed is here:

https://github.com/pombase/genome_changelog/blob/master/results/genome_sequence_changes.tsv

If we don't include this in the table, then I can definitely create a help section explaining this.

**Comment by kimrutherford** [comment]
> Maybe I would rename the block to "Gene Structure History", to be a bit more specific.

That makes sense.  I'll make that change.

**Comment by kimrutherford** [comment]
> Maybe I would rename the block to "Gene Structure History", to be a bit more specific.

Done!

**Comment by manulera** [comment]
TODO @manulera document section and refer to old assembly sequences

**Comment by manulera** [comment]
Hi @kimrutherford I have created a documentation page for the history section on this new branch https://github.com/pombase/website/compare/master...issue_1993

I think you have to create a component in Angular for this, but if I can do anything else on my side let me know.

Also, I was thinking that now that the svn files are accessible, we could add an extra field on the table with a link to the file of that genome version. I could add this as an extra field to the table. Links would look like this:

https://curation.pombase.org/pombe-embl-repo/trunk/pMIT.contig?p=5949

**Comment by kimrutherford** [comment]
>  I have created a documentation page for the history section on this new branch

Thanks!  I'm going to merge the change so we can see how it looks.

> I think you have to create a component in Angular for this, but if I can do anything else on my side let me know.

I won't need to add a component.  It's all (sort of) automatic.  To add documentation we just need to add a `.md` file in the documentation directory.

> Also, I was thinking that now that the svn files are accessible, we could add an extra field on the table with a link to the file of that genome version. I could add this as an extra field to the table.

Good plan!

**Comment by kimrutherford** [comment]
>  I'm going to merge the change so we can see how it looks.

This is how it looks on my desktop:
https://desktop.kmr.nz/documentation/gene-page-gene-structure-history

**Comment by manulera** [comment]
> Also, I was thinking that now that the svn files are accessible, we could add an extra field on the table with a link to the file of that genome version. I could add this as an extra field to the table.

Ok, I have added this. An extra column in the usual file (`results/all_previous_coords.tsv`) called `genome_snapshot`, link to the svn or ftp folder of the revision prior to the introduction of the change.

**Comment by manulera** [comment]
I guess once that's in we can close this one.

**Comment by kimrutherford** [comment]
> Also, I was thinking that now that the svn files are accessible, we could add an extra field on the table with a link to the file of that genome version. 

I'm adding this now.  I'm not sure what text to use for the column header and the text to use for the links.  This is what I currently have:

![image](https://user-images.githubusercontent.com/90474/224256298-7b08538f-f2c0-4b4d-b48f-5ad24c35b86a.png)

**Comment by manulera** [comment]
I think Genome snapshot is good for the header, perhaps for the value we can use the revision number, could be something like:

* svn:8760 for svn revisions
* ftp:20110304 for ftp revisions

That constitutes some sort of unique identifier.

**Comment by kimrutherford** [comment]
Hi @manulera

> I think Genome snapshot is good for the header, perhaps for the value we can use the revision number, could be something like:
>
>    svn:8760 for svn revisions
>    ftp:20110304 for ftp revisions

That makes sense.  I've implemented that: https://www.pombase.org/gene/SPCC622.17

I've noticed a small issue.  The SVN URLs before revision 374 need to be slightly different.  I got the paths wrong when I initially set up SVN so I had to move things in changeset 374.  Would you be able to change the URLs for revisions 1 to 373 in the data file to this pattern: `https://curation.pombase.org/pombe-embl-repo?p=REVISION`?  So drop the "/trunk" from the URL for the old revisions.

Thanks!

**Comment by manulera** [comment]
Hi @kimrutherford, the svn urls are now fixed with https://github.com/pombase/genome_changelog/commit/662bc55643c663b8c2975dc8ed1dd5f9e13fd999

**Comment by kimrutherford** [comment]
> Hi @kimrutherford, the svn urls are now fixed

Thanks!
